numpy/core/numeric.py

```diff
@@ -458,10 +465,31 @@ class ndarray is returned.
     >>> np.asanyarray(a) is a
     True
 
+    Make sure the result is compatible with a specified dtype
+
+    >>> a = np.ones(2, dtype=np.int16)
+    >>> np.asarray(a, precision=np.float32)
+    array([ 1.,  1.], dtype=float32)
+    >>> a = np.ones(2, dtype=np.complex128)
+    >>> np.asarray(a, precision=np.float32)
+    array([ 1.+0.j,  1.+0.j])
+    >>> a = np.ones(2, dtype=np.complex64)
+    >>> np.asarray(a, precision=0.0)
+    array([ 1.+0.j,  1.+0.j], dtype=complex64)
+
     """
-    return array(a, dtype, copy=False, order=order)
+    if precision is not None and dtype is not None:
+        raise ValueError('cannot specify both dtype and precision')
+
+    ret = array(a, dtype, copy=False, order=order)
```

Comment:
I don't think `np.can_cast` does the right thing. `result_type` returns the smallest dtype to which all the arguments can be promoted. For instance
```
In [1]: for dt in typecodes['AllInteger']:
   ...:     dold = dtype(dt)
   ...:     dnew = result_type(dold, float32)
   ...:     print "%s -> %s" % (dold, dnew)
   ...: 
int8 -> float32
uint8 -> float32
int16 -> float32
uint16 -> float32
int32 -> float64
uint32 -> float64
int64 -> float64
uint64 -> float64
int64 -> float64
uint64 -> float64
int64 -> float64
uint64 -> float64
```